evm_arithmetization/src/generation/prover_input.rs

```diff
@@ -74,9 +86,40 @@ impl<F: Field> GenerationState<F> {
     fn run_trie_ptr(&mut self, input_fn: &ProverInputFn) -> Result<U256, ProgramError> {
         let trie = input_fn.0[1].as_str();
         match trie {
-            "state" => Ok(U256::from(self.trie_root_ptrs.state_root_ptr)),
+            "state" => match self.trie_root_ptrs.state_root_ptr {
+                Some(state_root_ptr) => Ok(state_root_ptr),
+                None => {
+                    self.set_preinit = true;
+                    let mut new_content = self.memory.get_preinit_memory(Segment::TrieData);
+
+                    let n = load_state_mpt(&self.inputs.trimmed_tries, &mut new_content)?;
+
+                    self.memory.insert_preinitialized_segment(
+                        Segment::TrieData,
+                        crate::witness::memory::MemorySegmentState {
+                            content: new_content,
+                        },
+                    );
+                    Ok(n)
+                }
+            }
```

Comment:
Mested `match` constructs are not the easiest to read.  The pattern here seems to apply: [`map_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#handling-a-result-based-fallback)

Reply:
[Overview](https://doc.rust-lang.org/std/option/index.html#transforming-contained-values)